packages/core/contracts/optimistic-oracle-v2/implementation/ManagedOptimisticOracleV2.sol

```diff
@@ -0,0 +1,419 @@
+        address requester,
+        bytes32 identifier,
+        uint256 timestamp,
+        bytes memory ancillaryData,
+        uint256 customLiveness
+    ) external nonReentrant() onlyRequestManager() {
+        require(
+            _getState(requester, identifier, timestamp, ancillaryData) == State.Requested,
+            "setCustomLiveness: Requested"
+        );
+        _validateLiveness(customLiveness);
+        _getRequest(requester, identifier, timestamp, ancillaryData).requestSettings.customLiveness = customLiveness;
+    }
+
+    /**
+     * @notice Sets the proposer whitelist for a request.
+     * @dev This can also be set in advance of the request as the timestamp is omitted from the mapping key derivation.
+     * @param requester sender of the initial price request.
+     * @param identifier price identifier to identify the existing request.
+     * @param ancillaryData ancillary data of the price being requested.
+     * @param whitelist address of the whitelist to set.
+     */
+    function requestManagerSetProposerWhitelist(
+        address requester,
+        bytes32 identifier,
+        bytes memory ancillaryData,
+        address whitelist
+    ) external nonReentrant() onlyRequestManager() {
+        bytes32 requestId = _getId(requester, identifier, 0, ancillaryData);
+        customProposerWhitelists[requestId] = DisableableAddressWhitelistInterface(whitelist);
+        emit CustomProposerWhitelistSet(requestId, requester, identifier, ancillaryData, whitelist);
+    }
+
+    /**
+     * @notice Proposes a price value on another address' behalf. Note: this address will receive any rewards that come
+     * from this proposal. However, any bonds are pulled from the caller.
+     * @dev Timestamp is omitted from the whitelist key derivation, so it would also apply for repeated requests.
+     * @param proposer address to set as the proposer.
+     * @param requester sender of the initial price request.
+     * @param identifier price identifier to identify the existing request.
+     * @param timestamp timestamp to identify the existing request.
+     * @param ancillaryData ancillary data of the price being requested.
+     * @param proposedPrice price being proposed.
+     * @return totalBond the amount that's pulled from the caller's wallet as a bond. The bond will be returned to
+     * the proposer once settled if the proposal is correct.
+     */
+    function proposePriceFor(
+        address proposer,
+        address requester,
+        bytes32 identifier,
+        uint256 timestamp,
+        bytes memory ancillaryData,
+        int256 proposedPrice
+    ) public override returns (uint256 totalBond) {
+        DisableableAddressWhitelistInterface whitelist =
+            _getEffectiveProposerWhitelist(requester, identifier, ancillaryData);
+
+        require(whitelist.isOnWhitelist(proposer), "Proposer not whitelisted");
+        require(whitelist.isOnWhitelist(msg.sender), "Sender not whitelisted");
```

Comment:
Question: Do we expect msg.sender to ever differ from the proposer here?
I would have assumed `msg.sender == proposer`, effectively making `proposePriceFor` behave like `proposePrice` conceptually. So we should only need to check that `msg.sender` is whitelisted.

Reply:
yes, it might not be strictly necessary to check for `proposer`, but since the `msg.sender` is not emitted in events it would be harder to curate the whitelist in case the admin needs to remove proposers with high dispute rate. Otherwise a `msg.sender` that is part of whitelist could submit bad proposals using different proposer address making it harder for the manager to analyze. Its still possible to track the `msg.sender`, but that would require inspecting transaction call traces that puts extra burden on maintainer.